### ISSUE TITLE
Update MacOS Agent instaler (MeshAgent.mpkg)

### ISSUE THREAD

**Issue by Mantatools** [Ylianst/MeshCentral#3397]
Hello,
we find out that we are not able to do anything on connected Apple devices. We just see the desktop, we are able to move with mice but we can not click on anything and we can not see any open windows. 
Terminal a and files work properly.  
We are using macOS Monterey v.12.1
Can I ask you to check it? 

Thank you!

**Comment by krayon007** [comment]
Did you enable the permissions requested on macos? There are two separate ones. If you only accept one of the two, I think is when you get the scenario where you only see the background.

**Comment by Mantatools** [comment]
Yes, I have allowed meshagent_osx64 in Security & Privacy - Accessibility and Screen recording. Should I have it also somewhere else? Thanks

**Comment by Mantatools** [comment]
And I'm still getting this message.

![photo_2022-01-04_10-26-26](https://user-images.githubusercontent.com/91133890/148037750-ecd35d1f-1825-4c92-a953-484131cf07e1.jpg)

I found that someone install Agent via wget but it also doesn't work
https://issueexplorer.com/issue/Ylianst/MeshAgent/104

**Comment by Mantatools** [comment]
@krayon007 @Ylianst Any updates here? Thank you!

**Comment by krayon007** [comment]
I haven't reproduced it, but I will try a clean install with my Monterey System. Let me get back to you.

**Comment by I2700I** [comment]
Same issue with the M1 Mac, not sure if the architecture is relevant.

**Comment by JSuenram** [comment]
Same here Intel MAC with Monterey 12 and 12.1. On connect system is always asking for Screenrecord but it is already granted. Just Desktop-Background and Mousemove, no Windows at all.

**Comment by oezh** [comment]
Same here! permissions already granted but keeps asking en every connection

> Same here Intel MAC with Monterey 12 and 12.1. On connect system is always asking for Screenrecord but it is already granted. Just Desktop-Background and Mousemove, no Windows at all.

**Comment by krayon007** [comment]
I'll take a look on my Monterey System. One thing I've seen on occasion, is that the gatekeeper gets confused with permissions, as I've noticed that the UI is keyed by app name, but the actual permission is keyed by file hash. Anyways, I've seen scenarios where the permissions are messed up until you delete all the permissions for the app, and let it re-request permissions

In this case, it's not the agent causing the issues, but the OS itself that is getting in a wierd state

**Comment by zackrspv** [comment]
Is there an update on this? 3/22/2022 on M1 MacBook Pros' still doing the same thing. Even though permissions are granted.

**Comment by krayon007** [comment]
Have you tried removing all permissions, and letting it re-request the permission?

**Comment by I2700I** [comment]
> Have you tried removing all permissions, and letting it re-request the permission?

Yep. Two computers for test purposes, same issue with both.

**Comment by Mantatools** [comment]
@Ylianst @krayon007 
Any updates here?
Thank you!

**Comment by matt-cylanda** [comment]
Hey all,

I was about to make a new ticket about this, but found this. I've been experiencing similar issues. On some it's the above where there's just the desktop and we can control the mouse, but can't see anything regardless of how we enable permissions.

On my new 14" Macbook Pro, I'm only seeing a black screen in MeshCentral and do not have the ability to move the mouse. Please let us know if we can forward any log data.

**Comment by krayon007** [comment]
Have you guys tried downloading the binary installer and using that to install the agent, instead of the package installer? Does that make a difference for you?

**Comment by matt-cylanda** [comment]
Hi Bryan,

I have. On MacOS 11, this seemed to fix my issue on my M1 Macbook Air. Once I replaced the universal binary with the Arm one, things began working, but so far on both Intel and Arm I haven't been able to MeshCentral to work with Monterey.

On a side note, we have a handful of clients running Mac, and Monterey in general has been REALLY difficult to deploy things like screen sharing on. The solution for some apps has been going into system recovery and disabling SIP.

Actually now that it crossed my mind, I'm going to try it now to see if it fixes this issue.

**Comment by matt-cylanda** [comment]
Just to update, disabling the SIP system did not help. I can also confirm this is an issue on both M1 and Intel Macs, the issue seems to be MacOS 12 Monterey. I've also been experiencing LOTS of other super random issues with this OS release, it's so far the buggiest Mac OS release I've ever seen.

**Comment by matt-cylanda** [comment]
Just as an update, I went ahead and disabled the SIP system on my 14" M1 and it was a no go. I also removed the agent from permissions, rebooted and tried again with no luck.

Also, I can confirm, this is not an architecture issue. We can't get MeshAgent working on any Intel Macs with Monterey either. And as a side note to alleviate some of the "this program is broken" complaining... In all my experience using Macs for the past 20 years, this has got to be the buggiest Mac OS release I've ever seen, LOTS of random issues all over the place.

**Comment by matt-cylanda** [comment]
There might be an architecture thing after all. I accidentally downloaded the x86 installer and ran it on my system. This time the permissions work correctly, and I can see the main desktop (not my other screens), but I can't control anything.

For me so far, the universal and arm binaries just give a black screen, and the permissions alert always prompts to enable permissions for screen recording.

**Comment by matt-cylanda** [comment]
Ok... I can definitely confirm this is a binary related issue. At least on my M1, the only agent that works is the x86 binary, the Universal and Arm binaries both endlessly ask for permission each time I try to connect.

**Comment by barart** [comment]
Same here on Monterey on a M1... how to reinstall/update the agent with the arm binary? can somebody point on this?

**Comment by wizdude** [comment]
I've got the same issue here. A brand new Macbook Air 2020 M1 running the latest Monterey 12.5 and installed the "Apple macOS Agent (64bit)". continuous popups asking for screen record permissions.

@matt-cylanda you made reference to the x86 binary. what am I selecting to get this? If I select "Linux / BSD / macOS Binary Installer" the only Apple download i can see is "Apple OSX Universal" which i think you mentioned does not work.

**Comment by jbsolutios** [comment]
Exactly the same problem here.

I have a MacBook Pro 16 M1 running Monterey 12.5. 

The normal "Apple macOS" client installs fine, but every time you try to use Desktop > Connect, you are prompted with this and it never works:

<img width="573" alt="Screenshot 2022-07-28 at 23 05 24" src="https://user-images.githubusercontent.com/25282890/181649077-90f5c676-27bd-4e49-aa9b-67cbaecc2641.png">

Using the Apple OSX Universal client works, but it's no where near as elegant.

We have a large user base of M1 Macs and I can see that this is going to be a major issue if we try to move them to MeshCentral @krayon007

**Comment by krayon007** [comment]
So if I understand correctly, when using an M1 Mac you are only seeing issues with the arm binary but not the universal or Intel binary?

**Comment by jbsolutios** [comment]
When using this one (macOS Agent (64bit), remote control does not work. Each time you are prompted to grant access.

<img width="400" alt="Screenshot 2022-07-28 at 23 54 18" src="https://user-images.githubusercontent.com/25282890/181650665-03be05b1-02f1-423a-9db7-33d3be68d4b2.png">

When using this one, everything works, however it just installs wherever you have downloaded it. I cannot find any information on how to get this to work in the same was the the macOS Agent (64bit). 

<img width="404" alt="Screenshot 2022-07-28 at 23 54 52" src="https://user-images.githubusercontent.com/25282890/181650882-301c9b2e-ab97-4ac8-a9c4-78074ef52d6e.png">

**Comment by jbsolutios** [comment]
The ideal for me @krayon007 would be for the macOS Agent (64bit) to be updated so that it works correctly, as a .pkg is much neater for mass deployment.

**Comment by krayon007** [comment]
Interesting. I may have to update my M1 to Monterey
 Mine is still on Big Sur and works fine with all the binaries.

**Comment by jbsolutios** [comment]
That would be really helpful if you could, as we are looking to roll this out for all of our clients, but a lot of them are using MacBook Pro M1 with Monterey.


Regards


John Benson
MANAGING DIRECTOR

+ Solutios Ltd - 38 Skyline Court - Grange Yard - London - SE1 3AN - United Kingdom
t: + 44 (0) 20 7167 5555 - e: ***@***.***
________________________________

**Comment by barart** [comment]
Keep in mind that Apple Macs with M1 chip are not 64bits, they are M1 (Apple Silicon) an ARM architecture, macOS agent 64bits is not going to work on M1 macs as they are different architecture...

**Comment by jbsolutios** [comment]
Yes, but it does work when Rosetta2 is installed, with the exception of remote control.

Would an Apple Silicon build of the .pkg be a possibility please?


Regards


John Benson
MANAGING DIRECTOR

+ Solutios Ltd - 38 Skyline Court - Grange Yard - London - SE1 3AN - United Kingdom
t: + 44 (0) 20 7167 5555 - e: ***@***.***
________________________________

**Comment by si458** [comment]
hey gang, just looking at all the mac issues now as somebody kindly donated an M1 to fix bugs!
has anybody found a way to actually get the remote control working on M1 12.7.2 ?
ive tried the mpkg and the binarys and they just dont wanna work?
if i can find a way to get it working i can look into why it needs to be done that way

**Comment by b-hal** [comment]
Yes, after a lot of try I got it to work with macOS 13 Ventura. 
macOS 12 Monterey will be EOL end-of-life at the end of 2024,  so testing with Ventura makes more sense.

**Comment by si458** [comment]
i think ive found the issue, it appears the universal binarys arent playing nicely with screen recording.

so ive added the option under the 'binary install' so you can download either universal, x64, or arm binarys now for osx!
if you download the arm binary and install, it works no problems!

ive still not quite got the mpkg working yet but baby steps...

p.s yes i have rosetta installed cuz i installed the xcode-select and developer tools which needed it installed anyways

**Comment by si458** [comment]
update. ive found a work around fix for others if intrested in trying
install the binary as universal, then BEFORE you connect to the device the first time, run `agentupdate` from the console tab
this will force the meshagent to redownload the arm version and use that instead of the universal one
then when you connect, it will ask for screen record permission, so give it and it works!
you sadly still have to disconnect and then reconnect for it to ask for control permission too,
but working on a fix for that!
